fsm_codegen/src/codegen.rs

```diff
@@ -240,16 +278,25 @@ pub fn build_state_transitions(fsm: &FsmDescription) -> quote::Tokens {
                 let mut state_exit = quote! {
                     self.inspection.on_state_exit(&current_state, &event_ctx).await;
 
-                    if let Err(error) = self.states.#source_state_field.on_exit(&mut event_ctx).await {
-                        break Err(error);
+                    {
+                        let state = self.states.#source_state_field.read().await;
+                        if let Err(error) = state.on_exit(&mut event_ctx).await {
+                            break Err(error);
+                        }
                     }
                 };
 
                 let mut state_entry = quote! {
-                    self.inspection.on_state_entry(&#states_ty::#target_state, &event_ctx).await;
-
-                    if let Err(error) = self.states.#target_state_field.on_entry(&mut event_ctx).await {
-                        break Err(error);
+                    self.inspection.on_state_entry(
+                        &#states_ty::#target_state(std::sync::Arc::clone(&self.states.#target_state_field)),
```

Comment:
self.states.#target_state_field.clone() werkt niet?

Reply:
To ensure that no expensive clones happen if code changes due to refactors. With this code, it's guaranteed that only an Arc clone happens.